### PR TITLE
Testing: Enable passing directory root to test runner

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to enhance the test runner by enabling a developer to pass a directory root for which tests should be sought and run.

__Testing instructions:__

Existing test behavior should remain unaffected. Specifically, run variations of test commands, including the entire test suite, test suites specific to directories (`client`, `server`, `test`), and specifying a single file.

Next, ensure that you can also use the test runner to run tests within a specific directory. For example:

```
npm run test-client client/state/post-types
```

...should run all test files in `client/state/post-types`, including those in `client/state/post-types/test` and in `client/state/post-types/taxonomies/test`

/cc @gziolo @blowery

### CODE REVIEW

**Comment by blowery** [comment]
LGTM. @gziolo ?

**Comment by gziolo** [comment]
That was my hunch, it should be relatively easy to add this feature ;) Nicely done!

I found two edge cases which we could fix later, one existing and one new:
* `npm run test-client server/api/` - it tries to load the tests from the server folder when you use client runner, it was also the case before for single file/
* `npm run test-client client/state/post-types/test/` - doesn't run any tests, when you skip test folder it runs tests from all subfolders, so it isn't possible to limit to current folder

Don't treat it as a blocker :) I'm giving 👍  and you can 🚢

File: test/runner.js [hunk 1]
```diff
@@ -48,13 +48,21 @@ if ( process.env.CIRCLECI ) {
...
 mocha.suite.afterAll( boot.after );
 
-files = program.args;
-if ( files.length === 0 ) {
-	files = glob.sync( process.env.TEST_ROOT + '/**/test/*.@(js|jsx)' );
-	if ( program.nodeTotal > 1 ) {
-		files = files.filter( ( file, index ) => index % program.nodeTotal === program.nodeIndex );
+files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
+files = files.reduce( ( memo, filePath ) => {
+	if ( /\.jsx?$/i.test( filePath ) ) {
```
Comment (gziolo) [inline test/runner.js]: I took another look, so we could warn when a file doesn't begin with `process.env.TEST_ROOT`.

File: test/runner.js [hunk 2]
```diff
@@ -48,13 +48,21 @@ if ( process.env.CIRCLECI ) {
...
+		return memo.concat( filePath );
 	}
+
+	return memo.concat( glob.sync( filePath + '/**/test/*.@(js|jsx)' ) );
```
Comment (gziolo) [inline test/runner.js]: Also here, we could add condition which adds different suffix when `filePath` ends with `/test`.

**Comment by aduth** [comment]
@gziolo : Great suggestions. I implemented both in f8d3607 and e4d0b39. Let me know what you think.

File: test/runner.js [hunk 3]
```diff
@@ -48,13 +50,38 @@ if ( process.env.CIRCLECI ) {
...
-	if ( program.nodeTotal > 1 ) {
-		files = files.filter( ( file, index ) => index % program.nodeTotal === program.nodeIndex );
+files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
+files = files.reduce( ( memo, filePath ) => {
+	// Validate test root matches specified file paths
+	if ( ! filePath.startsWith( process.env.TEST_ROOT ) ) {
+		console.log(
+			chalk.red.bold( 'WARNING:' ),
+			chalk.yellow( 'Invalid arguments passed to test runner (files with mismatched test root)' )
+		);
```
Comment (gziolo) [inline test/runner.js]: We probably should return `memo` here.

**Comment by aduth** [comment]
@gziolo 

>We probably should return `memo` here.

Yeah, you're probably right. Updated to prevent the invalid arguments from being included in the files tested in 26b7574 and improved the messaging a bit.

image

File: test/runner.js [hunk 4]
```diff
@@ -48,13 +50,41 @@ if ( process.env.CIRCLECI ) {
...
-		files = files.filter( ( file, index ) => index % program.nodeTotal === program.nodeIndex );
+files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
+files = files.reduce( ( memo, filePath ) => {
+	// Validate test root matches specified file paths
+	if ( ! filePath.startsWith( process.env.TEST_ROOT ) ) {
+		console.warn(
+			chalk.red.bold( 'WARNING:' ),
+			chalk.yellow( 'Invalid argument passed to test runner. Paths must match test root `' + process.env.TEST_ROOT + '`.' )
+		);
+		console.warn( ' - ' + filePath + '\n' );
```
Comment (gziolo) [inline test/runner.js]: It's looking sweet right now :)

File: test/runner.js [hunk 5]
```diff
@@ -48,13 +50,41 @@ if ( process.env.CIRCLECI ) {
 mocha.suite.beforeAll( boot.before );
 mocha.suite.afterAll( boot.after );
 
-files = program.args;
-if ( files.length === 0 ) {
-	files = glob.sync( process.env.TEST_ROOT + '/**/test/*.@(js|jsx)' );
-	if ( program.nodeTotal > 1 ) {
-		files = files.filter( ( file, index ) => index % program.nodeTotal === program.nodeIndex );
+files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
+files = files.reduce( ( memo, filePath ) => {
```
Comment (gziolo) [inline test/runner.js]: We could refactor a bit this code now. It grew quite big recently. What do you think about moving that logic inside one of dependent file and using something similar to `flowRight` which would call functions like:

```js
executeUntil( validateTestRoot, appendIndividualFile, appendDiscoveredFiles );
```

Then we would have code chunks grouped by unique use cases and we don't need to add inline comments, because functions names are self-explanatory.

File: test/README.md [hunk 6]
```diff
@@ -33,6 +33,9 @@ Example for client:
 > npm run test-server server/config/test/parser.js
 > # run single test suite from test folder
 > npm run test-test test/helpers/use-nock/test/index.js
+> # run test suite for all files in a specific folder
+> npm run test-client client/state
+> npm run test-client client/state/posts/test
```
Comment (gziolo) [inline test/README.md]: Documentation update, nice 👍

File: test/runner.js [hunk 7]
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env node
+'use strict'; // eslint-disable-line strict
```
Comment (gziolo) [inline test/runner.js]: Node 6.x should help us :)

Comment (aduth) [inline test/runner.js]: >Node 6.x should help us :)

Does 6.x allow using block-scope variables without strict mode? That's the issue I was running in to.

File: test/runner.js (hunk 5 above)
Comment (aduth) [inline test/runner.js]: >We could refactor a bit this code now.

I like the idea of self-documenting function names, but I'll probably hold off on this 'til a future pull request.

File: test/runner.js (hunk 7 above)
Comment (gziolo) [inline test/runner.js]: It should. See http://node.green/. They get to 96% coverage of ES6 features :)

File: test/runner.js (hunk 5 above)
Comment (gziolo) [inline test/runner.js]: Fair enough, this PR is really great improvement anyway 👍

**Comment by gziolo** [comment]
I testes it with very different path combinations and it works like a chars. Thanks for working on that one. This should be very useful 💯 

Some combinations I used:
* `npm run test-client client/state/post-types client/state/current-user/test/actions.js` - it allows to combine folders and files :)
* `npm run test-server server/devdocs/ client/state/current-user/` - it executes only server related tests and warns about the other path.
* `run test-client client/state/post-types/test/ client/state/post-types/` - it doesn't run duplicates!

Let's merge it asap 🎯

**Comment by aduth** [comment]
>it doesn't run duplicates!

I _think_ it's probably not smart enough to dedupe the files on its own, but I'm guessing Mocha is handling this on our behalf.
